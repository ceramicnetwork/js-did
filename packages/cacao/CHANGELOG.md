# @didtools/cacao

## 3.0.0

### Major Changes

- Upgrade to multiformats 13 and related packages impacted

### Patch Changes

- Updated dependencies
  - @didtools/codecs@2.0.0
  - @didtools/siwx@2.0.0

## 2.1.0

### Minor Changes

- c64ce06: Light-weight parsing of SIWx message: reduces size of bundle

### Patch Changes

- Updated dependencies [c64ce06]
  - @didtools/siwx@1.0.0

## 2.0.1

### Patch Changes

- Add new codecs package
- Updated dependencies
  - @didtools/codecs@1.0.1
