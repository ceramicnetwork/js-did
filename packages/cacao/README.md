# CACAO

A library to represent chain-agnostic Object Capabilities (OCAP), created using [EIP-4361](https://github.com/ethereum/EIPs/blob/5e9b0fe0728e160f56dd1e4cbf7dc0a0b1772f82/EIPS/eip-4361.md) (or similar for other blockchains), as an [IPLD](https://ipld.io/) object.

## Examples

### Convert between CACAO and SIWE (EIP-4361)

```typescript
const siweMessage = new SiweMessage({
  domain: 'service.org',
  address: address,
  statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
  uri: 'did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8',
  version: '1',
  nonce: '32891757',
  issuedAt: '2021-09-30T16:25:24.000Z',
  chainId: '1',
  resources: [
    'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
    'https://example.com/my-web2-claim.json',
    'ceramic://k2t6wyfsu4pg040dpjpbla1ybxof65baldb7fvmeam4m3n71q0w1nslz609u2d',
  ],
})

const cacao = Cacao.fromSiweMessage(siweMessage)
const siweMessage2 = SiweMessage.fromCacao(cacao)
```

### Creating and signing a CACAO with private-key Wallet

```typescript
import { Wallet } from '@ethersproject/wallet'
import { Cacao, CacaoBlock, SiweMessage } from 'ceramic-cacao'

const wallet = Wallet.createRandom()
const address = wallet.address

const siweMessage = new SiweMessage({
  domain: 'service.org',
  address: address,
  statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
  uri: 'did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8',
  version: '1',
  nonce: '32891757',
  issuedAt: '2021-09-30T16:25:24.000Z',
  chainId: '1',
  resources: [
    'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
    'https://example.com/my-web2-claim.json',
    'ceramic://k2t6wyfsu4pg040dpjpbla1ybxof65baldb7fvmeam4m3n71q0w1nslz609u2d',
  ],
})
const signature = await wallet.signMessage(siweMessage.toMessage())
siweMessage.signature = signature

const cacao = Cacao.fromSiweMessage(siweMessage)
```

### Usage with `EthereumAuthProvider` `requestCapability` to update a `TileDocument`

```typescript
import { Web3Provider } from "@ethersproject/providers"
import { EthereumAuthProvider } from "@ceramicnetwork/blockchain-utils-linking";
import { Ed25519Provider } from "key-did-provider-ed25519"
import { DID } from "dids"
import * as KeyDidResolver from "key-did-resolver";
import { TileDocument } from "@ceramicnetwork/stream-tile";
import { CeramicClient } from "@ceramicnetwork/http-client";
import type { Cacao } from "ceramic-cacao"

// Create the EthereumAuthProvider
const web3Provider = new Web3Provider(...) // connect to a provider
const address = "0xAB..." // get the signer address
const ethereumAuthProvider = new EthereumAuthProvider(web3Provider.provider, address) // Note: we pass the underlying RPC provider, not the ethers.js wrapped version

// Create a determinstic document for the user
const ceramic = new CeramicClient(CERAMIC_API_URL) // Ceramic HTTP Client
const deterministicDocument = await TileDocument.deterministic(ceramic, {
    deterministic: true,
    family: 'randomFamily',
    controllers: [`did:pkh:eip155:1:${address}`]
})

// Create a session key for the dApp
const seed = ... // 32 bytes of entropy
const didProvider = new Ed25519Provider(seed)
const didKey = new DID({
    provider: didProvider,
    resolver: KeyDidResolver.getResolver()
})
await didKey.authenticate()

// Request capability from user
const cacao = await ethereumAuthProvider.requestCapability(didKey.id, [deterministicDocument.id.toUrl()])

// Attach capability to session key
const didKeyWithCap = didKey.withCapability(cacao);
await didKeyWithCap.authenticate()

// Update user's TileDocument using session key that has the capability
await deterministicDocument.update({ foo: 'bar' }, {}, {
    asDid: didKeyWithCap
})
```

## License

Dual licensed with APACHE and MIT
