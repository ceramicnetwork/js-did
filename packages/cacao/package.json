{
  "name": "ceramic-cacao",
  "version": "1.4.0",
  "description": "Typescript library for Ceramic OCAP",
  "author": "Haardik <hhaardik@uwaterloo.ca>",
  "license": "(Apache-2.0 OR MIT)",
  "type": "module",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": "./dist/index.js"
  },
  "files": [
    "dist"
  ],
  "engines": {
    "node": ">=14.14"
  },
  "sideEffects": false,
  "scripts": {
    "build:clean": "del dist",
    "build:js": "swc src -d ./dist --config-file ../../.swcrc",
    "build:types": "tsc --emitDeclarationOnly --skipLibCheck",
    "build": "pnpm run build:clean && pnpm run build:types && pnpm run build:js",
    "lint": "eslint src test --fix",
    "test": "node --experimental-vm-modules ../../node_modules/jest/bin/jest.js",
    "test:ci": "pnpm run test --ci --coverage",
    "prepare": "pnpm run build",
    "prepublishOnly": "package-check"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ceramicnetwork/js-did.git"
  },
  "keywords": [
    "DID",
    "identity",
    "OCAP",
    "capabilities"
  ],
  "bugs": {
    "url": "https://github.com/ceramicnetwork/js-did/issues"
  },
  "homepage": "https://github.com/ceramicnetwork/js-did#readme",
  "dependencies": {
    "@ethersproject/wallet": "^5.5.0",
    "@ipld/dag-cbor": "^7.0.1",
    "@stablelib/ed25519": "^1.0.2",
    "apg-js": "^4.1.1",
    "caip": "^1.0.0",
    "multiformats": "^9.5.1",
    "uint8arrays": "^3.0.0"
  },
  "devDependencies": {
    "@types/luxon": "^3.0.0",
    "luxon": "^3.0.1"
  }
}
