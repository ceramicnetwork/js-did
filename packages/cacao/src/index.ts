export * from './siwx/siwe.js'
export * from './siwx/siws.js'
export * from './siwx/siwx.js'
export * from './cacao.js'
