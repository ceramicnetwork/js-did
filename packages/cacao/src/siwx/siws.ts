import { SignatureType, SiwxMessage } from './siwx.js'

export class SiwsMessage extends SiwxMessage {
  toMessage(): string {
    return super.toMessage('Solana')
  }

  signMessage(): Uint8Array {
    let message: Uint8Array
    switch (this.type) {
      case SignatureType.PERSONAL_SIGNATURE: {
        message = new TextEncoder().encode(this.toMessage())
        break
      }

      default: {
        message = new TextEncoder().encode(this.toMessage())
        break
      }
    }

    return message
  }
}
