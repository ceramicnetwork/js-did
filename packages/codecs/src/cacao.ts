import { array, keyof, optional, sparse, strict, string } from 'codeco'

export const CacaoHeader = strict({ t: keyof({ eip4361: null, caip122: null }) }, 'CacaoHeader')
export type CacaoHeader = { t: 'eip4361' | 'caip122' }

export const CacaoPayload = sparse(
  {
    domain: string,
    iss: string,
    aud: string,
    version: string,
    nonce: string,
    iat: string,
    nbf: optional(string),
    exp: optional(string),
    statement: optional(string),
    requestId: optional(string),
    resources: optional(array(string)),
  },
  'CacaoPayload'
)
export type CacaoPayload = {
  domain: string
  iss: string
  aud: string
  version: string
  nonce: string
  iat: string
  nbf?: string
  exp?: string
  statement?: string
  requestId?: string
  resources?: Array<string>
}

export const CacaoSignature = strict(
  {
    t: keyof({
      eip191: null,
      eip1271: null,
      'solana:ed25519': null,
      'tezos:ed25519': null,
      'stacks:secp256k1': null,
    }),
    s: string,
  },
  'CacaoSignature'
)
export type CacaoSignature = {
  t: 'eip191' | 'eip1271' | 'solana:ed25519' | 'tezos:ed25519' | 'stacks:secp256k1' | 'webauthn:p256'
  s: string,
  aad?: Uint8Array
}

export const Cacao = sparse(
  { h: CacaoHeader, p: CacaoPayload, s: optional(CacaoSignature) },
  'Cacao'
)
export type Cacao = {
  h: CacaoHeader
  p: CacaoPayload
  s?: CacaoSignature
}

export const SignedCacao = strict(
  { h: CacaoHeader, p: CacaoPayload, s: CacaoSignature },
  'SignedCacao'
)
export type SignedCacao = {
  h: CacaoHeader
  p: CacaoPayload
  s: CacaoSignature
}
