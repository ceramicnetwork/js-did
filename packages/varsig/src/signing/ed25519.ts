import { MAGIC } from '../magic.js'
import { SigningAlgo } from '../signing.js'
import { ed25519 } from '@noble/curves/ed25519'
import type { BytesTape } from '../bytes-tape.js'

const SIGIL = MAGIC.ED25519

// @ts-ignore
function prepareVerifier(tape: BytesTape): SigningAlgo {
  return {
    kind: SIGIL,
    // eslint-disable-next-line @typescript-eslint/require-await
    verify: async (input, signature, verificationKey): Promise<boolean> => {
      return ed25519.verify(signature, input, verificationKey)
    },
  }
}

function readSignature(tape: BytesTape): Uint8Array {
  return tape.read(64)
}

export const ED25519 = { SIGIL, prepareVerifier, readSignature }
