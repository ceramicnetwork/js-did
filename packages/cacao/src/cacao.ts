// @ts-nocheck
import { verifyMessage } from '@ethersproject/wallet'
import * as dagCbor from '@ipld/dag-cbor'
import { verify } from '@stablelib/ed25519'
import { AccountId } from 'caip'
import * as multiformats from 'multiformats'
import * as Block from 'multiformats/block'
import { sha256 as hasher } from 'multiformats/hashes/sha2'
import { fromString as u8aFromString } from 'uint8arrays/from-string'
import { SiweMessage } from './siwx/siwe.js'
import { SiwsMessage } from './siwx/siws.js'
import { asLegacyChainIdString } from './siwx/siwx.js'

// 5 minute default clockskew
const CLOCK_SKEW_DEFAULT_SEC = 5 * 60

// CACAOs issued after that day must be of new format
export const LEGACY_CHAIN_ID_REORG_DATE = new Date('2022-09-20').valueOf()

export type Header = {
  t: 'eip4361' | 'caip122'
}

export type Payload = {
  domain: string
  iss: string
  aud: string
  version: string
  nonce: string
  iat: string
  nbf?: string
  exp?: string
  statement?: string
  requestId?: string
  resources?: Array<string>
}

export type Signature = {
  t: 'eip191' | 'eip1271' | 'solana:ed25519'
  s: string
}
export type Cacao = {
  h: Header
  p: Payload
  s?: Signature
}

export type VerifyOptions = {
  /**
   * @param atTime - the point in time the capability is being verified for
   */
  atTime?: Date
  /**
   * @param expPhaseOutSecs - Number of seconds that a capability stays valid for after it was expired
   */
  revocationPhaseOutSecs?: number
  /**
   * @param clockSkewSecs - Number of seconds of clock tolerance when verifying iat, nbf, and exp
   */
  clockSkewSecs?: number

  /**
   * @param disableExpirationCheck - Do not verify expiration time
   */
  disableExpirationCheck?: boolean
}

export namespace Cacao {
  export function fromSiweMessage(siweMessage: SiweMessage): Cacao {
    const cacao: Cacao = {
      h: {
        t: 'eip4361', // TODO: Replace with "caip122" after 2022-09-24
      },
      p: {
        domain: siweMessage.domain,
        iat: siweMessage.issuedAt,
        iss: `did:pkh:eip155:${siweMessage.chainId}:${siweMessage.address}`,
        aud: siweMessage.uri,
        version: siweMessage.version,
        nonce: siweMessage.nonce,
      },
    }

    if (siweMessage.signature) {
      cacao.s = {
        t: 'eip191',
        s: siweMessage.signature,
      }
    }

    if (siweMessage.notBefore) {
      cacao.p.nbf = siweMessage.notBefore
    }

    if (siweMessage.expirationTime) {
      cacao.p.exp = siweMessage.expirationTime
    }

    if (siweMessage.statement) {
      cacao.p.statement = siweMessage.statement
    }

    if (siweMessage.requestId) {
      cacao.p.requestId = siweMessage.requestId
    }

    if (siweMessage.resources) {
      cacao.p.resources = siweMessage.resources
    }

    return cacao
  }

  export function fromSiwsMessage(siwsMessage: SiwsMessage): Cacao {
    const cacao: Cacao = {
      h: {
        t: 'caip122',
      },
      p: {
        domain: siwsMessage.domain,
        iat: siwsMessage.issuedAt,
        iss: `did:pkh:solana:${siwsMessage.chainId}:${siwsMessage.address}`,
        aud: siwsMessage.uri,
        version: siwsMessage.version,
        nonce: siwsMessage.nonce,
      },
    }

    if (siwsMessage.signature) {
      cacao.s = {
        // https://github.com/solana-labs/wallet-adapter/issues/179
        t: 'solana:ed25519',
        s: siwsMessage.signature,
      }
    }

    if (siwsMessage.notBefore) {
      cacao.p.nbf = siwsMessage.notBefore
    }

    if (siwsMessage.expirationTime) {
      cacao.p.exp = siwsMessage.expirationTime
    }

    if (siwsMessage.statement) {
      cacao.p.statement = siwsMessage.statement
    }

    if (siwsMessage.requestId) {
      cacao.p.requestId = siwsMessage.requestId
    }

    if (siwsMessage.resources) {
      cacao.p.resources = siwsMessage.resources
    }

    return cacao
  }

  export async function fromBlockBytes(bytes: Uint8Array): Promise<Cacao> {
    const block = await Block.decode({ bytes: bytes, codec: dagCbor, hasher: hasher })
    return block.value as Cacao
  }

  export function verify(cacao: Cacao, options: VerifyOptions = {}): void {
    if (cacao.s?.t === 'eip191') {
      return verifyEIP191Signature(cacao, options)
    } else if (cacao.s?.t === 'solana:ed25519') {
      return verifySolanaSignature(cacao, options)
    }
    throw new Error('Unsupported CACAO signature type')
  }

  export function verifyEIP191Signature(cacao: Cacao, options: VerifyOptions) {
    if (!cacao.s) {
      throw new Error(`CACAO does not have a signature`)
    }

    const atTime = options.atTime ? options.atTime.getTime() : Date.now()
    const clockSkew = (options.clockSkewSecs ?? CLOCK_SKEW_DEFAULT_SEC) * 1000

    if (
      Date.parse(cacao.p.iat) > atTime + clockSkew ||
      Date.parse(cacao.p.nbf) > atTime + clockSkew
    ) {
      throw new Error(`CACAO is not valid yet`)
    }

    const phaseOutMS = options.revocationPhaseOutSecs ? options.revocationPhaseOutSecs * 1000 : 0

    if (
      !options.disableExpirationCheck &&
      Date.parse(cacao.p.exp) + phaseOutMS + clockSkew < atTime
    ) {
      throw new Error(`CACAO has expired`)
    }

    const recoveredAddress = verifyMessage(
      SiweMessage.fromCacao(cacao).toMessage(),
      cacao.s.s
    ).toLowerCase()
    const recoveredAddresses = [recoveredAddress]

    if (Date.parse(cacao.p.iat) <= LEGACY_CHAIN_ID_REORG_DATE) {
      const legacyChainIdRecoveredAddress = verifyMessage(
        asLegacyChainIdString(SiweMessage.fromCacao(cacao), 'Ethereum'),
        cacao.s.s
      ).toLowerCase()
      recoveredAddresses.push(legacyChainIdRecoveredAddress)
    }

    const issuerAddress = AccountId.parse(cacao.p.iss.replace('did:pkh:', '')).address.toLowerCase()
    if (!recoveredAddresses.includes(issuerAddress)) {
      throw new Error(`Signature does not belong to issuer`)
    }
  }
}

export function verifySolanaSignature(cacao: Cacao, options: VerifyOptions) {
  if (!cacao.s) {
    throw new Error(`CACAO does not have a signature`)
  }
  const atTime = options.atTime ? options.atTime.getTime() : Date.now()

  if (Date.parse(cacao.p.iat) > atTime || Date.parse(cacao.p.nbf) > atTime) {
    throw new Error(`CACAO is not valid yet`)
  }

  const phaseOutMS = options.revocationPhaseOutSecs ? options.revocationPhaseOutSecs * 1000 : 0

  if (!options.disableExpirationCheck && Date.parse(cacao.p.exp) + phaseOutMS < atTime) {
    throw new Error(`CACAO has expired`)
  }

  const msg = SiwsMessage.fromCacao(cacao)
  const sig = cacao.s.s

  const messageU8 = msg.signMessage()
  const sigU8 = u8aFromString(sig, 'base58btc')
  const issAddress = AccountId.parse(cacao.p.iss.replace('did:pkh:', '')).address
  const pubKeyU8 = u8aFromString(issAddress, 'base58btc')

  if (!verify(pubKeyU8, messageU8, sigU8)) {
    throw new Error(`Signature does not belong to issuer`)
  }
}

export type CacaoBlock = {
  value: Cacao
  cid: multiformats.CID
  bytes: Uint8Array
}

export namespace CacaoBlock {
  export function fromCacao(cacao: Cacao): Promise<CacaoBlock> {
    return Block.encode<Cacao, number, number>({
      value: cacao,
      codec: dagCbor,
      hasher: hasher,
    })
  }
}
