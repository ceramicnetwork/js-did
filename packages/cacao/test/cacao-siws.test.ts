import { extractPublicKeyFromSecretKey, sign } from '@stablelib/ed25519'
import { fromString } from 'uint8arrays/from-string'
import { toString } from 'uint8arrays/to-string'
import { Cacao, CacaoBlock } from '../src/cacao.js'
import { SiwsMessage } from '../src/siwx/siws.js'
describe('Cacao SIWS', () => {
  const solanaSecretKey = fromString(
    '92e08e39aee87d53fe263913bf9df6615c1c909860a1d3ad57bd0e6e2e507161ecbf1e2d9da80d3ae09de54ce71cbff723e291e7a4b133ce10993be5edfaca50',
    'hex'
  )

  test('Can create and verify Cacao Block for Solana', async () => {
    const msg = new SiwsMessage({
      domain: 'service.org',
      address: toString(extractPublicKeyFromSecretKey(solanaSecretKey), 'base58btc'),
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8',
      version: '1',
      nonce: '32891757',
      issuedAt: '2021-09-30T16:25:24.000Z',
      chainId: '1',
      resources: [
        'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
        'https://example.com/my-web2-claim.json',
      ],
    })

    const signData = msg.signMessage()
    const rawSignature = sign(solanaSecretKey, signData)
    const signature = toString(rawSignature, 'base58btc')
    msg.signature = signature

    const cacao = Cacao.fromSiwsMessage(msg)
    const block = await CacaoBlock.fromCacao(cacao)
    expect(block).toMatchSnapshot()
    expect(() => Cacao.verify(cacao)).not.toThrow()
  })

  test('Converts between Cacao and SiwsMessage', () => {
    const msg = new SiwsMessage({
      domain: 'service.org',
      address: toString(extractPublicKeyFromSecretKey(solanaSecretKey), 'base58btc'),
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'https://service.org/login',
      version: '1',
      nonce: '32891757',
      issuedAt: '2021-09-30T16:25:24.000Z',
      chainId: '1',
      resources: [
        'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
        'https://example.com/my-web2-claim.json',
      ],
    })

    const cacao = Cacao.fromSiwsMessage(msg)
    const siws = SiwsMessage.fromCacao(cacao)
    expect(siws).toEqual(msg)
  })

  test('ok after exp if within phase out period', () => {
    const fixedDate = new Date('2021-10-14T07:18:41Z')
    const msg = new SiwsMessage({
      domain: 'service.org',
      address: toString(extractPublicKeyFromSecretKey(solanaSecretKey), 'base58btc'),
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'https://service.org/login',
      version: '1',
      nonce: '32891757',
      issuedAt: fixedDate.toISOString(),
      expirationTime: new Date(fixedDate.valueOf() + 5 * 1000).toISOString(),
      chainId: '1',
      resources: [
        'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
        'https://example.com/my-web2-claim.json',
      ],
    })

    const signData = msg.signMessage()
    const rawSignature = sign(solanaSecretKey, signData)
    const signature = toString(rawSignature, 'base58btc')
    msg.signature = signature

    const cacao = Cacao.fromSiwsMessage(msg)
    const expiredTime = new Date(fixedDate.valueOf() + 10 * 1000)
    expect(() =>
      Cacao.verify(cacao, { atTime: expiredTime, revocationPhaseOutSecs: 20 })
    ).not.toThrow()
  })

  test('fail after exp if after phase out period', () => {
    const fixedDate = new Date('2021-10-14T07:18:41Z')
    const msg = new SiwsMessage({
      domain: 'service.org',
      address: toString(extractPublicKeyFromSecretKey(solanaSecretKey), 'base58btc'),
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'https://service.org/login',
      version: '1',
      nonce: '32891757',
      issuedAt: fixedDate.toISOString(),
      expirationTime: new Date(fixedDate.valueOf() + 5 * 1000).toISOString(),
      chainId: '1',
      resources: [
        'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
        'https://example.com/my-web2-claim.json',
      ],
    })

    const signData = msg.signMessage()
    const rawSignature = sign(solanaSecretKey, signData)
    const signature = toString(rawSignature, 'base58btc')
    msg.signature = signature

    const cacao = Cacao.fromSiwsMessage(msg)
    const expiredTime = new Date(fixedDate.valueOf() + 10 * 1000)
    expect(() => Cacao.verify(cacao, { atTime: expiredTime, revocationPhaseOutSecs: 1 })).toThrow(
      `CACAO has expired`
    )
  })
})
