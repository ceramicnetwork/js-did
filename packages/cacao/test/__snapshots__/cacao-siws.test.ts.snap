// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cacao SIWS Can create and verify Cacao Block for Solana 1`] = `
Block {
  "asBlock": [Circular],
  "bytes": Uint8Array [
    163,
    97,
    104,
    161,
    97,
    116,
    103,
    99,
    97,
    105,
    112,
    49,
    50,
    50,
    97,
    112,
    168,
    99,
    97,
    117,
    100,
    120,
    56,
    100,
    105,
    100,
    58,
    107,
    101,
    121,
    58,
    122,
    54,
    77,
    107,
    114,
    66,
    100,
    78,
    100,
    119,
    85,
    80,
    110,
    88,
    68,
    86,
    68,
    49,
    68,
    67,
    120,
    101,
    100,
    122,
    86,
    86,
    66,
    112,
    97,
    71,
    105,
    56,
    97,
    83,
    109,
    111,
    88,
    70,
    65,
    101,
    75,
    78,
    103,
    116,
    65,
    101,
    114,
    56,
    99,
    105,
    97,
    116,
    120,
    24,
    50,
    48,
    50,
    49,
    45,
    48,
    57,
    45,
    51,
    48,
    84,
    49,
    54,
    58,
    50,
    53,
    58,
    50,
    52,
    46,
    48,
    48,
    48,
    90,
    99,
    105,
    115,
    115,
    120,
    61,
    100,
    105,
    100,
    58,
    112,
    107,
    104,
    58,
    115,
    111,
    108,
    97,
    110,
    97,
    58,
    49,
    58,
    71,
    119,
    65,
    70,
    52,
    53,
    122,
    106,
    102,
    121,
    71,
    122,
    85,
    98,
    100,
    51,
    105,
    51,
    104,
    88,
    120,
    122,
    71,
    101,
    117,
    99,
    104,
    122,
    69,
    90,
    88,
    119,
    112,
    82,
    89,
    72,
    90,
    77,
    53,
    57,
    49,
    50,
    70,
    49,
    101,
    110,
    111,
    110,
    99,
    101,
    104,
    51,
    50,
    56,
    57,
    49,
    55,
    53,
    55,
    102,
    100,
    111,
    109,
    97,
    105,
    110,
    107,
    115,
    101,
    114,
    118,
    105,
    99,
    101,
    46,
    111,
    114,
    103,
    103,
    118,
    101,
    114,
    115,
    105,
    111,
    110,
    97,
    49,
    105,
    114,
    101,
    115,
    111,
    117,
    114,
    99,
    101,
    115,
    130,
    120,
    53,
    105,
    112,
    102,
    115,
    58,
    47,
    47,
    81,
    109,
    101,
    55,
    115,
    115,
    51,
    65,
    82,
    86,
    103,
    120,
    118,
    54,
    114,
    88,
    113,
    86,
    80,
    105,
    105,
    107,
    77,
    74,
    56,
    117,
    50,
    78,
    76,
    103,
    109,
    103,
    115,
    122,
    103,
    49,
    51,
    112,
    89,
    114,
    68,
    75,
    69,
    111,
    105,
    117,
    120,
    38,
    104,
    116,
    116,
    112,
    115,
    58,
    47,
    47,
    101,
    120,
    97,
    109,
    112,
    108,
    101,
    46,
    99,
    111,
    109,
    47,
    109,
    121,
    45,
    119,
    101,
    98,
    50,
    45,
    99,
    108,
    97,
    105,
    109,
    46,
    106,
    115,
    111,
    110,
    105,
    115,
    116,
    97,
    116,
    101,
    109,
    101,
    110,
    116,
    120,
    65,
    73,
    32,
    97,
    99,
    99,
    101,
    112,
    116,
    32,
    116,
    104,
    101,
    32,
    83,
    101,
    114,
    118,
    105,
    99,
    101,
    79,
    114,
    103,
    32,
    84,
    101,
    114,
    109,
    115,
    32,
    111,
    102,
    32,
    83,
    101,
    114,
    118,
    105,
    99,
    101,
    58,
    32,
    104,
    116,
    116,
    112,
    115,
    58,
    47,
    47,
    115,
    101,
    114,
    118,
    105,
    99,
    101,
    46,
    111,
    114,
    103,
    47,
    116,
    111,
    115,
    97,
    115,
    162,
    97,
    115,
    120,
    87,
    52,
    75,
    77,
    80,
    90,
    106,
    112,
    83,
    99,
    118,
    54,
    99,
    75,
    112,
    52,
    76,
    49,
    113,
    101,
    85,
    99,
    119,
    113,
    97,
    89,
    76,
    50,
    89,
    86,
    97,
    109,
    105,
    81,
    114,
    75,
    53,
    100,
    69,
    121,
    84,
    69,
    98,
    74,
    56,
    81,
    86,
    114,
    113,
    112,
    116,
    72,
    81,
    70,
    68,
    51,
    118,
    99,
    84,
    110,
    78,
    75,
    105,
    72,
    69,
    78,
    107,
    111,
    114,
    115,
    121,
    111,
    120,
    97,
    70,
    71,
    121,
    111,
    99,
    72,
    75,
    89,
    83,
    110,
    103,
    71,
    97,
    106,
    97,
    116,
    110,
    115,
    111,
    108,
    97,
    110,
    97,
    58,
    101,
    100,
    50,
    53,
    53,
    49,
    57,
  ],
  "cid": Object {
    "code": 113,
    "hash": Uint8Array [
      18,
      32,
      134,
      156,
      89,
      50,
      0,
      205,
      20,
      139,
      179,
      35,
      130,
      149,
      165,
      183,
      186,
      84,
      135,
      170,
      250,
      175,
      58,
      146,
      94,
      19,
      235,
      84,
      187,
      151,
      148,
      127,
      150,
      244,
    ],
    "version": 1,
  },
  "value": Object {
    "h": Object {
      "t": "caip122",
    },
    "p": Object {
      "aud": "did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8",
      "domain": "service.org",
      "iat": "2021-09-30T16:25:24.000Z",
      "iss": "did:pkh:solana:1:GwAF45zjfyGzUbd3i3hXxzGeuchzEZXwpRYHZM5912F1",
      "nonce": "32891757",
      "resources": Array [
        "ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu",
        "https://example.com/my-web2-claim.json",
      ],
      "statement": "I accept the ServiceOrg Terms of Service: https://service.org/tos",
      "version": "1",
    },
    "s": Object {
      "s": "4KMPZjpScv6cKp4L1qeUcwqaYL2YVamiQrK5dEyTEbJ8QVrqptHQFD3vcTnNKiHENkorsyoxaFGyocHKYSngGaj",
      "t": "solana:ed25519",
    },
  },
}
`;
